### PR TITLE
Add tests to ensure post controller accepts the same values it returns when creating a post

### PR DESCRIPTION
**Pull request by adamsilverstein**
- Adds some tests to ensure API accepts null date data and empty slug data for draft posts.
- Adds a test to ensure API accepts back all values it returns when creating a new post as valid.
- Adds an is_object test to slug validation in `prepare_item_for_database`, preventing empty slug objects values from being added to the post

### CODE REVIEW

**Comment by adamsilverstein** [comment]
Fixes https://github.com/WP-API/WP-API/issues/2696

File: lib/endpoints/class-wp-rest-posts-controller.php [hunk 1]
```diff
@@ -797,7 +797,7 @@ protected function prepare_item_for_database( $request ) {
 			}
 		}
 		// Post slug.
-		if ( ! empty( $schema['properties']['slug'] ) && isset( $request['slug'] ) ) {
+		if ( ! empty( $schema['properties']['slug'] ) && isset( $request['slug'] ) && ! empty( $request['slug'] ) && ! is_object( $request['slug'] ) ) {
```
Comment (adamsilverstein) [inline lib/endpoints/class-wp-rest-posts-controller.php]: A blank slug sent to the api winds up here as an empty object which was passing the empty test, I didn’t figure out why. Without this change, the `test_update_draft_post_with_returned_draft_post_data` test below fails.

Comment (joehoyle) [reply lib/endpoints/class-wp-rest-posts-controller.php]: Hmm @adamsilverstein did you figure out where this empty object is coming from?

Comment (adamsilverstein) [reply lib/endpoints/class-wp-rest-posts-controller.php]: Not yet, I'll dig into it @joehoyle. I would expect a blank string here.

Comment (adamsilverstein) [reply lib/endpoints/class-wp-rest-posts-controller.php]: @rmccue @joehoyle I tracked this down to our use of the `sanitize_title` to sanitize the slugs. `sanitize_title` takes an optional second parameter as a 'fallback'. in class-wp-rest-controller:L807 in `sanitize_params` where we use `call_user_func`, we are passing the value, then also passing the request object and the argument key.

In this case `sanitize_title` sees an empty string as failing and returns the fallback, the request object, and the post update fails.

Do you know why the request and key are added to the sanitization callback? can those be removed? Going to dig into when they were added.

https://core.trac.wordpress.org/browser/trunk/src/wp-includes/rest-api/class-wp-rest-request.php#L807

Comment (adamsilverstein) [reply lib/endpoints/class-wp-rest-posts-controller.php]: ps. I removed the is_object test and now my `test_update_draft_post_with_empty_slug` test fails

**Comment by adamsilverstein** [comment]
@joehoyle I think we need to change our sanitize callback to fix this issue.

In https://github.com/WP-API/WP-API/pull/2763/commits/a3c95b5884c3d22344233da8758a61e17248c1b2 I added a new `rest_sanitize_sanitization callback for the slug that ignores the passed $request and $key, calling`sanitize_title`on the value alone, and properly returning a blank string when a blank string is sanitized. I was mixed about whether this should be named as is or`rest_sanitize_title`

File: plugin.php [hunk 2]
```diff
@@ -487,3 +487,16 @@ function rest_is_boolean( $maybe_bool ) {
 		return false;
 	}
 }
+
+/**
```
Comment (joehoyle) [inline plugin.php]: Hrmmm I feel like we already added this somewhere, @rmccue is that true?

Comment (rmccue) [reply plugin.php]: Indeed; it's in the base controller as `WP_REST_Controller->sanitize_slug`

Comment (rmccue) [reply plugin.php]: Which, in fact, is already used above, so not sure why this PR is changing it?

Comment (adamsilverstein) [reply plugin.php]: @joehoyle @rmccue I think you were one step ahead of me, I just updated develop and see the new function. updating the PR, all I have to add now is the new tests, everything checks out with this fix in place.

**Comment by adamsilverstein** [comment]
After updating develop, everything checks out for this issue. I'm leaving the test in place for consideration because they validate that everything works as expected.

**Comment by rmccue** [comment]
@adamsilverstein The tests are failing here; does this need `develop` merged in again to resolve?

**Comment by adamsilverstein** [comment]
@rmccue not sure its my branch? I can refresh patch if needed, I'm seeing this in Travis failure: `PHP Fatal error:  Cannot use object of type stdClass as array in /home/travis/build/WP-API/WP-API/lib/fields/class-wp-rest-meta-fields.php on line 104`

**Pull request closed without merging**